Add delete_bkg command - fix #891

DougBurke opened the pull request:
# Summary

Add the delete_bkg call. Fix #891.

# Details

This originally had a lot of clean-up documention/typing work which got moved into #2124, hence some of the dicsussion below in the comments does not match the current code.

We add

- `delete_bkg` call to the UI layer
- `get_backgrounds` to the `DataPHA` object, in an attempt to "better type" the code (e.g. to avoid some cases where we know a value is not None because the identifier exists but the code does not)

One question that came up from this work was
- #2113

which I do not plan to address here.

# Issues to resolve

- this code follows the existing `delete_data` behavior and does not error out if a dataset has no backgrounds (it's a little more complicated than this, but this is the main issue) and the question is should it error out
- our code essentially mimics a dictionary (since it does) so maybe we should make this the primary API rather than existing calls
- the current code actually lets you (I **think**) temporarily remove a background but leave it still associated with the parent `DataPHA` whereas the new code does not; this strikes me as a subtle disaster waiting to strike

# Example

```
>>> from sherpa.astro import ui
WARNING: imaging routines will not be available, 
failed to import sherpa.image.ds9_backend due to 
'RuntimeErr: DS9Win unusable: Could not find ds9 on your PATH'
WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available
>>> ui.load_pha("sherpa-test-data/sherpatest/3c273.pi")
WARNING: systematic errors were not found in file 'sherpa-test-data/sherpatest/3c273.pi'
statistical errors were found in file 'sherpa-test-data/sherpatest/3c273.pi'
but not used; to use them, re-read with use_errors=True
read ARF file sherpa-test-data/sherpatest/3c273.arf
read RMF file sherpa-test-data/sherpatest/3c273.rmf
WARNING: systematic errors were not found in file 'sherpa-test-data/sherpatest/3c273_bg.pi'
statistical errors were found in file 'sherpa-test-data/sherpatest/3c273_bg.pi'
but not used; to use them, re-read with use_errors=True
read background file sherpa-test-data/sherpatest/3c273_bg.pi
>>> ui.subtract()
```

So, what happens if we delete the background?

```
>>> ui.delete_bkg()
dataset 1: background subtraction has been removed
```

And we can no-longer get it:

```
>>> ui.get_bkg()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 1, in get_bkg
  File "/home/dburke/sherpa/sherpa-conda/sherpa/astro/ui/utils.py", line 7010, in get_bkg
    raise IdentifierErr('getitem', 'background data set',
sherpa.utils.err.IdentifierErr: background data set 1 in PHA data set 1 has not been set
```

DougBurke committed:
ui: improve documentation of delete_data

Note that it does nothing if the given identifier does not exist.

DougBurke committed:
data: add explicit checks that background datasets exist

For several get_background calls we assumed they would always
pass; not, that may be the case (from the context they are
run in) but we can not guarantee that and the code is labelled
as maybe returning a None, so we handle these cases by

- skipping the code
- or error-ing out.

DougBurke committed:
UI: add delete_bkg call fix #891

This call is modelled on delete_data but

- we do not explicitly store the associated backgrounds as we
  rely on the DataPHA object to manage that

- there is a choice of when to do nothing or to error out, so
  we do nothing if there is no dataset, or there is a PHA dataset
  but no associated background, but we do error out if the dataset
  is not a PHA one.

hamogu commented on sherpa/astro/data.py:
```diff
@@ -4645,7 +4700,10 @@ def notice_response(self, notice_resp=True, noticed_chans=None):
             _notice_resp(noticed_chans, arf, rmf)
 
-    def notice(self, lo=None, hi=None, ignore=False, bkg_id=None):
+    def notice(self, lo=None, hi=None,
```
Is there a reason that there is no type hint for "lo" and "hi"? 
If we type a function, I would prefer to type all parameters (except in special cases where it's just too hard to do)

hamogu commented on sherpa/astro/data.py:
```diff
@@ -4807,6 +4865,10 @@ def notice(self, lo=None, hi=None, ignore=False, bkg_id=None):
         #
         for bid in bkg_ids:
             bkg = self.get_background(bid)
+            if bkg is None:
```
Maybe this is OK, but it look odd to me. `bkg_ids` is returned from `self.background_ids` and I would expect every background to have an identifier. Or is there a situation where that's not the case?`

hamogu replied:
Or is this for the case when no background it set at all? But then the comment in the line below is confusing.

DougBurke replied:
So this is a case where I claim our API doesn't help us. We should have a way to be able to loop through the backgrounds, which would mean we don't need the `if bkg is None` statement to please the type checker

At the moment, since `get_background` returns `DataPHA | None` . the type checker has no way to know that `get_background` can not return `None` here.

I'm not sure the best way to square the circle

- live with the old code, and annoy the type checker gods
- annoy readers with pointless code
- address the API issues, which is just not a good thing to be doing at this stage of the release cycle (well, not when I have other changes I want to get in ;-)

hamogu replied:
For some reason, I've never managed to get a type checker to work with my editor. I know many people do and it should work, but something is off and I never bothered to spend the time to track it down. So I'm fine with annoying the type checker gods ... but you do much more work in the codebase than I do and the typecheckers did help us (more precisely: you) find a few useful things in the past. So, if you prefer "pointless code" that's also fine with me - just click "resolve the conversation"

hamogu commented on sherpa/astro/ui/tests/test_astro_ui_unit.py:
```diff
@@ -3168,6 +3168,117 @@ def test_set_bkg_not_a_background(clean_astro_ui):
+@pytest.mark.parametrize("id", [None, 1, "x"])
+@pytest.mark.parametrize("bkg_id", [None, 1, "y"])
+def test_delete_bkg_no_bkg(id, bkg_id, clean_astro_ui):
+    """Check we can delete a background when nothing exists.
```
I'm not happy with this behaviour. Do we need it for backwards compatibility? I'm worried about silent errors, e.g. I have several datasets and I want to remove the background for one of them. I mistype the id, and don't get an error, so I assume it all worked, when in fact it didn't do anything. 
I would prefer to raise an error when a user tries to remove a background that does not exist.

DougBurke replied:
This mirrors `delete_data`:

```
>>> from sherpa.astro import ui
WARNING: imaging routines will not be available, 
failed to import sherpa.image.ds9_backend due to 
'RuntimeErr: DS9Win unusable: Could not find ds9 on your PATH' 
>>> ui.delete_data()
>>> ui.delete_data(1)
>>> ui.delete_data("x")
```

hamogu commented on sherpa/ui/utils.py:
```diff
@@ -4069,6 +4069,8 @@ def delete_data(self, id: Optional[IdType] = None) -> None:
         The source expression is not removed by this function.
 
+        The routine does nothing if the given dataset does not exist.
+
```
I know it's just a doc change, so this is not a request to change anything, but I general I think it's not good practice to fail silently.

hamogu commented on sherpa/astro/ui/utils.py:
```diff
@@ -174,19 +179,25 @@ def _check_pha_tabstops(data: DataPHA,
-                   get_err, colname):
+def _save_errorcol(session,
+                   idval: Optional[IdType],
+                   filename,
```
Do we support pathlib? If not (or we don't know for sure) we could just type it as string for now.

```suggestion
                   filename: str,
```

DougBurke replied:
So, general pathlib support is 

- #1070

and I'd hoped to get to it, but am not sure I will for 4.17.0.

However, this is covered by a combination of 

- https://github.com/sherpa/sherpa/pull/2114#discussion_r1732743640
- https://github.com/sherpa/sherpa/pull/2114#discussion_r1732781845

hamogu commented on sherpa/astro/ui/utils.py:
```diff
@@ -654,7 +677,12 @@ def _get_show_bkg_source(self, id=None, bkg_id=None):
+    def show_bkg(self,
+                 id: Optional[IdType] = None,
+                 bkg_id: Optional[IdType] = None,
+                 outfile=None,
```
```suggestion
                 outfile: str = None,
```

hamogu replied:
Same below. I'm sure @DougBurke had a good reason to not type this, so I want to discuss here instead of making suggestions for all of them at this point.

DougBurke replied:
Now, I did work on this in a recent PR 

- #2054 
 
 and came up with

https://github.com/sherpa/sherpa/blob/307ddbbd4b7ff98e3c9720915f05aa530f4a1262/sherpa/fit.py#L570-L579

so maybe I can use that.

DougBurke replied:
But not in this PR.

hamogu replied:
OK.

hamogu commented on sherpa/astro/ui/utils.py:
```diff
@@ -2525,7 +2577,9 @@ def load_filter(self, id, filename=None, bkg_id=None, ignore=False,
     # talks about 2 cols, but experimentation suggests 1 col.
     #
-    def load_grouping(self, id, filename=None, bkg_id=None, *args, **kwargs):
+    def load_grouping(self, id, filename=None,
```
Inconsisteny with the doctring, which says filename is a str.

hamogu replied:
Same happens below in different locations where the other parameters of a function are typed in this PR. Again, I want to discuss here and not mark all instances at this stage.

DougBurke replied:
I point your attention to the "Notes" section of this (and many other routines in this module):

https://github.com/sherpa/sherpa/blob/307ddbbd4b7ff98e3c9720915f05aa530f4a1262/sherpa/astro/ui/utils.py#L2528-L2576

I have been worrying about how to type these, or whether we even can, for a long time now. However, that was before I knew about [typing.overload](https://docs.python.org/3/library/typing.html#overload). I now think we could use that approach, but

- I have not tied it
- it would be massively invasive, so I don't think now is the time
- it looks like the semantics of `overload` is not entirely resolved - see the [discourse discussion](https://discuss.python.org/t/draft-typing-spec-chapter-for-overloads/60788) - which is always fun to see (although I'm not sure we are likely to hit on these areas)

hamogu replied:
Oh, I hate these "designed for easy interactive use". It's easy - unless you want to see what you need to put in. But we're not going to redesign the UI because I don't like it.
OK, I'll live with it.

hamogu commented on sherpa/astro/ui/utils.py:
```diff
@@ -3154,7 +3227,9 @@ def set_exposure(self, id, exptime=None, bkg_id=None):
         d.exposure = exptime
 
-    def set_backscal(self, id, backscale=None, bkg_id=None):
+    def set_backscal(self, id, backscale=None,
```
Is there a reason to just partially type this function?

hamogu replied:
Same below, but again, I'm not marking all of them.

DougBurke replied:
Because I do not know the full types that should be used (is it `float`, is it `float | Sequence[float]`, is it `float |  Sequence[float] | np.ndarray`, or ...). 

Typing  of these routines is "easier" once we have the classes better typed (since they define the types).

Here I was focussing on just adding types for "identifiers" (but, as noted above, I got a bit carried away). I htink I'm goong to remove the other types as they are causing too much discussion.

hamogu commented on sherpa/astro/ui/utils.py:
```diff
@@ -3491,7 +3580,10 @@ def get_error(self, id=None, filter=False, bkg_id=None):
+    def get_indep(self,
+                  id: Optional[IdType] = None,
+                  filter: bool = False,
+                  bkg_id: Optional[IdType] = None):
```
Should the return value be typed? (Same for a few other modified functions below)

DougBurke replied:
See https://github.com/sherpa/sherpa/pull/2114#discussion_r1732781845

hamogu commented on sherpa/astro/ui/utils.py:
```diff
@@ -6894,7 +7080,80 @@ def set_bkg(self, id, bkg=None, bkg_id=None):
+        The source expression for the background is not removed by
+        this function.
+
+        The routine does nothing if the given dataset or background
```
I would prefer to error out in both cases.

DougBurke replied:
I have a complex set of feelings around this

- I do appreciate the "error out" approach
- right up until I process data and write code where I have to explicitly catch the failure (or manually add validation checks) just because for this case the condition wasn't met. An example is calling "subtract" - sometimes it makes sense to error out and say "but you are already subtracted", but in other cases it can be okay to say "oh, you're already subtracted, I don't need to do anything".

I claim there's likely a range of approaches and it depends in part on whether (for Sherpa) it is a class (or low-level routine) or whether it's in the UI layer.
 
These are all generic points, and we need to think about the specifics of this. In this case I am just trying to follow the existing `delete_data` behaviour (kind of, as there are failure cases in the background case we don't have for the source case).

hamogu replied:
I believe I was the one who insisted the `subtract` should not give me an error if it's already subtracted because I run into cases where some of my datasets were already subtracted and some are not and then it failed in the middle (#766). So, point taken.

hamogu reviewed (changes requested):
Most of the changes are good. I have two major suggestions. I can be convinced otherwise, but I ask to consider both:

1. I think it's dangerous to fail silently, e.g. to not issue an error i I try to remove a background component that does not exist. I know some parts of Sherpa work like that and we don't want to break backwards compatibility, but I'm worried about e.g. `deletete_bgk('src', 'bgk1') not issuing an error. Users will assume that their background `bkg1` has been deleted, and not double check if they had a typo in the string.
2. Typing is optional and is a pain to do. Yet, if we type a function, I would prefer to type all parameters and the return value unless it's really hard (i.e. type at least all the things that the ints or strings).

DougBurke commented on sherpa/astro/data.py:
```diff
@@ -1556,10 +1566,10 @@ def _set_subtracted(self, val):
         return self._units
 
-    def _set_units(self, val):
+    def _set_units(self, val: str) -> None:
```
See https://github.com/sherpa/sherpa/pull/2114#discussion_r1732781845

DougBurke commented:
So, I'm basically yeeting all the "add types" comments into the void - or is it the abyss you yeet stuff into? - and will remove the extra types.

So, the main discussion is over whether something should be an error or ignored if "it isn't a valid operation".

DougBurke committed:
data: add a get_backgrounds method

This provides a more-useful API than the existing

    for bkg_id in foo.backround_ids:
        bkg = foo.get_background(bkg_id)
        # we know bkg is not None but the types do not ...

There are still places where background_ids is useful, and we
do not want to break existing code. There are some isses that
we *could* look at into how background_ids is created/stored/...
but I do not think it's worth it at this time.

DougBurke commented:
Rebased to address a simple conflict, from #2112, and added a commit to fix some of the typing rules I've added. There is also the addition of a `get_backgrounds` method which I added after a discussion with @hamogu over logic like

    for bkg_id in foo.background_ids:
        bkg = foo.get_background(bkg_id)
        # we know bkg is not None but the computer does not

so we can now replace it with

    for bkg_id, bkg in foo.get_backgrounds():
        # yay, we know bkg is a DataPHA

hamogu commented on sherpa/astro/data.py:
```diff
@@ -2716,12 +2721,31 @@ def get_background(self,
+        delete_background, get_background, set_background
+
+        """
+        return list(self._backgrounds.items())
```
I like this better than the previous use. However, I still wonder if this level of indirection is needed . This is a one-liner. Why not just use `data._background.items()` [note below] directly inline? (While they might be some special cases, where it needs to be used as a list, usually it's just iterated over and then `.items()` is fine - that's interable.

Obviously, this won't hurt performance or anything. I just personally find code that uses too many levels of indirection hard to read. I know that "In computer science, every problem can be solved with one more level of indirection" and there are valid use cases, but it seems overly complicated.

note: If that's the recommended patter, it should probably be made part of the official API and be called `backgrounds` instead of `_backgrounds`.

To be clear, I'm not going to veto this PR over this -- it's a valid design decision -- I'm just pointing out what I personally would prefer.

hamogu replied:
Thinking this further- do we need `self.background_ids`? Isn't that just `self._background.keys()`? Can be simplify a lot more by just making this dict non-private?

hamogu replied:
Last thought before I move to the next commit: If the pattern is just `data._background.items()` then we also don't need any extra tests that show that we really get back what we expect by calling the "get me the list of backgrounds" function because `data._background.items()`  is just Python and not Sherpa functionality - we can rely on the fact that `keys()` or `items()` works on a dict.

DougBurke replied:
So, `background_ids` is interestingly implemented, but it was designed a long time ago when we didn't have the certainty we now have (e.g. that ordering of keys is deterministic now). I wouldn't want to remove `background_ids` since it is an existing API and I don't see any benefit to removing it, but it could be cleaned up (but I'd rather do that in a later PR unless I gain a huge amount of enthusiasm and time).

For background information: this API was also designed to be somewhat language agnostic (as it had to allow working with S-Lang as well as Python) in which case having an explicit API rather than just relying on the Python dict interface was almost a necessity. We are obviously long gone from those requirements, but the old API still exists and I am not convinced of the use of replacing the API (as I don't want to break existing code, including within Sherpa itself), and is it worth having two APIs (explicit `get`/`set`/`delete` calls versus "use dictionary methods)?

Potential advantages to the indirection that the current design has is that we hide the internal implementation and we can check that the keys to the dict are IdType and the values are DataPHA objects when inserting them. We could instead overload the `backgrounds` property to allow this (but it gets a bit messy).

DougBurke replied:
So, with the current setup we can *technically* add a background to a dataset and then "remove" it - by setting the `background_ids` to an array without this identifier - but still have the background associated with the dataset.

I guess this would let you quickly answer the question "how much of a difference does background component xxx make", but I'm not sure it makes that much sense. This does, however, suggest that my `get_backgrounds` helper could break existing code (since it doesn't care about the `background_ids` field). It could be made to do so. Need to think.

hamogu replied:
We discussed this at the developer telecom and the comment "Need to think" make me think that it's not in my court, but waiting for @DougBurke to finish that thinking. But please let me know if I'm wrong - I don't want to hold thus up unnecessarily.

DougBurke commented:
Rebased now that #2124 has been merged.

hamogu commented:
There are so many open PRs that it's hard to keep track which one is in my court and which one is back on @DougBurke. Here, @DougBurke says ["need to think"](https://github.com/sherpa/sherpa/pull/2114#discussion_r1747044162) so I assume you are not waiting on me, right?

DougBurke commented:
yeah -with #2124 merged I am less gung-ho on pushing this one, so it's currently back-burnered on my set of burners. Which are burning.
